### PR TITLE
Go back to correct page in referral page

### PR DESCRIPTION
**Pull request by dukenv0307**
### Details

Go back to correct page in referral page
### Fixed Issues

$ https://github.com/Expensify/App/issues/31866
PROPOSAL: https://github.com/Expensify/App/issues/31866#issuecomment-1826380889

### Tests

1. Go to FAB > Start chat
2. Click on `Get $250`
3. Copy the URL
4. Send this URL to any report
5. Click on this URL in the message
6. Click on back button or confirm button in referral page
7. Verify that we go back to the report screen
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above
### QA Steps

1. Go to FAB > Start chat
2. Click on `Get $250`
3. Copy the URL
4. Send this URL to any report
5. Click on this URL in the message
6. Click on back button or confirm button in referral page
7. Verify that we go back to the report screen
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/129500732/6d77f3aa-adae-4c9e-b6bb-eeeb76daf583

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/129500732/6dd62a88-94f6-4505-8296-4649fa5de6ad

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/129500732/4093cf11-2064-497c-8748-c6d5f8d42ccb

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/129500732/b85c7ac9-2b40-4fc1-8795-1c543480855d

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/129500732/af2337ad-0be1-4c09-8ef9-98a217cddc4f

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/129500732/1bf34739-cb74-418c-ba06-5de144fe4556

</details>

### CODE REVIEW

File: src/pages/ReferralDetailsPage.js [hunk 1]
```diff
@@ -108,7 +93,7 @@ function ReferralDetailsPage({route, account}) {
                     success
                     style={[styles.w100]}
                     text={translate('common.buttonConfirm')}
-                    onPress={() => Navigation.goBack(getFallbackRoute())}
```
[the remark this answers is not in the document]
Comment (ArekChr) [reply src/pages/ReferralDetailsPage.js:111]: 1. Click on FAB > Request Money > switch to Manual
2. Enter the amount and click on Next
3. Click on 'Request money, get $250' to go to the referral page
4. Reload the page
5. Click on the 'Got it' Button

Expected Result:
Click on the 'Got it' Button takes you back to the Manual page

Actual Result:
Click on the 'Got it' Button takes you back to the main chat.

issue: https://github.com/Expensify/App/issues/31628

File: src/pages/ReferralDetailsPage.js [hunk 2]
```diff
@@ -54,11 +54,7 @@ function ReferralDetailsPage({route, account}) {
     const shouldShowBody2 = isShareCode;
     const shouldShowClipboard = contentType === CONST.REFERRAL_PROGRAM.CONTENT_TYPES.REFER_FRIEND || isShareCode;
 
-    function generateReferralURL(email) {
-        return `${CONST.REFERRAL_PROGRAM.LINK}/?thanks=${encodeURIComponent(email)}`;
-    }
-
-    function getFallbackRoute() {
+    const fallBackRoute = useMemo(() => {
```
Comment (ArekChr) [inline src/pages/ReferralDetailsPage.js:57]: We don't need memo optimization here, this is a cheap function

File: src/pages/ReferralDetailsPage.js [hunk 3]
```diff
@@ -67,6 +63,10 @@ function ReferralDetailsPage({route, account}) {
         };
 
         return fallbackRoutes[contentType];
+    }, [contentType]);
+
+    function generateReferralURL(email) {
```
Comment (ArekChr) [inline src/pages/ReferralDetailsPage.js:68]: Please bring back original order of functions

**Comment by ArekChr** [comment]
I'm still unclear. The back button "<" and the "got it" button seem not inconsistent but should be. Removing the fallback route for both buttons will resolve the issue of opening this page from the chat link. However, it will disrupt the usual flow after a page refresh, preventing us from returning to the search | start chat | sending money | etc., from a referral after reloading. I'm still contemplating the best solution, as the current setup needs to fulfil the requirements, we should rethink this flow

**Comment by ArekChr** [comment]
The issue is with how we are setting up the routes. Right now, when we go to 'request money', choose 'manual', and click 'referral', the route changes. It goes from this:

https://dev.new.expensify.com:8082/request/new/participants/

to this:

https://dev.new.expensify.com:8082/referral/request

If we merge all these routes into one, it will make the route history more transparent. This should solve both problems when refreshing the page, going back or pressing got it, and when opening the route from the chat link. 
After that, we won't need to store the fallback route page. The new, combined route would be:

https://dev.new.expensify.com:8082/request/new/participants/referral/request

@luacmartins, WDYT? I need your 👀 here.

**Comment by luacmartins** [comment]
Hmm that makes sense and would indeed make the routes more logical IMO. That being said, we need to account for the other content types - https://github.com/Expensify/App/blob/2e9cf17bfee27560f4a979fbf4815e3924f360c8/src/CONST.ts#L2829-L2833

**Comment by ArekChr** [comment]
Thanks, @luacmartins, for the insights. Let's go ahead with the implementation like we talked about. @dukenv0307, do you need additional information or clarification to continue working on this fix?

**Comment by ArekChr** [comment]
> @ArekChr So what is the expected behavior when we open referral page from the report message link and click on back button or got it button?

It depends on the link that is clicked. If the link includes only 'referral' as a subpath, it should lead back to the main page. If the 'referral' path includes additional segments like 'request/new/participants', it should navigate back to the specific page as indicated in the path. This navigation should be handled by React Router Dom.

**Comment by ArekChr** [comment]
@dukenv0307 Correct. Properly implementing routes is crucial for React Router to manage these scenarios effectively

@luacmartins, about the referral/request route – since it's just one page and not two, shouldn't we consider simplifying it to something like `referralRequest` for clarity and better route management?

**Comment by ArekChr** [comment]
> I think we can remove this route because after we change the route like the explanation above, no page redirect to this route.

Could you elaborate? Do you mean removing `referral` from url?

**Comment by ArekChr** [comment]
@dukenv0307 I think go back button have the same purpose like got it button

**Comment by ArekChr** [comment]
@luacmartins What your thoughts? Do we still need referral request under `new.expensify.com/create/request/participants/{transactionID}/{reportID}/referralRequest` url?

**Comment by luacmartins** [comment]
Yes, I think we do

**Comment by ArekChr** [comment]
@dukenv0307 After opening a link from report and going back from refferal page it shows not found page, please look into that
https://dev.new.expensify.com:8083/create/request/participants/1/8076487818448116/referral/request

Zrzut ekranu 2024-01-2 o 14 04 06

**Comment by ArekChr** [comment]
@dukenv0307 This is regression compared to staging and main branch. Steps to reproduce:

1. Go to manual request money
2. Input amount then press next
3. Go to referral page
4. Copy url and send it to any report
5. Click on sent URL 
6. Go back by pressing navigation chevron

**Comment by Beamanator** [comment]
Thanks for volunteering to take over review @situchan

**Comment by situchan** [comment]
@dukenv0307 please fix conflict. the branch is very behind

**Comment by situchan** [comment]
> @dukenv0307 please fix conflict. the branch is very behind

bump ^
6.4k commits are behind at this moment 😮

**Comment by situchan** [comment]
Please resolve conflict

**Comment by situchan** [comment]
Please merge main again. Lint is fixed just now in #36423

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by situchan** [comment]
Please pull main again. 1.4k commits are behind

**Comment by situchan** [comment]
@dukenv0307 please add context behind this change:

change

**Review by situchan — APPROVED** [review]
Latest changes LGTM

**Review by Beamanator — APPROVED** [review]

**Pull request merged**